builtin/game/item_s.lua

```diff
@@ -154,3 +156,47 @@ function core.strip_param2_color(param2, paramtype2)
 	-- paramtype2 == "color" requires no modification.
 	return param2
 end
+
+-- Content ID caching
+
+local old_get_content_id = core.get_content_id
+local old_get_name_from_content_id = core.get_name_from_content_id
+
+local name2content = setmetatable({}, {
+	__index = function(self, name)
+		return old_get_content_id(name)
```

Comment:
You could cache here already, to get some caching even before `builtin_shared.cache_content_ids` is called:
```suggestion
		local content_id = old_get_content_id(name)
		self[name] = content_id
		return content_id
```
And `builtin_shared.cache_content_ids` could then be simpler (i.e. just set the array part of content2name).
Like this you can also be sure not to forget any node, i.e. ignore.

Reply:
I'd rather not cache stuff while nodes are still being registered, as new aliases and stuff might invalidate the cache.

Reply:
New nodes shouldn't invalidate the cache, as they get new ids ([at least it seems so](https://github.com/minetest/minetest/blob/99c8295e71eb86333ae21b1083b6f55fb9479565/src/nodedef.cpp#L1197)).
But you're right. There's at least `remove_node` which can probably mess things up.

(The following could still be done to have a simpler `builtin_shared.cache_content_ids`:
```suggestion
		local content_id = old_get_content_id(name)
		if content_id_cache_on then
			self[name] = content_id
		end
		return content_id
```
But it's not too beneficial, so feel free to mark as resolved.)

Reply:
Complicating the hot path to simplify the init code doesn't really make sense.

Reply:
It's not hot if it's only called once... (It could get hot at init (when the `content_id_cache_on` I added there is false) though, yes.)

Reply:
IMO the information should all be cached at the beginning. If only some of the entries of `content2name` are filled in, they might be stored in the hash part of the table instead of the array part. This could be substantially slower.